src/libs/ReportActionsUtils.ts

```diff
@@ -961,6 +1021,98 @@ function getReportActionMessageText(reportAction: OnyxEntry<ReportAction> | Empt
+ * It determines the part of the message array to display ('visibleReportActions') based on the current linked message,
+ * and manages pagination through 'handleReportActionPagination' function.
+ *
+ * linkedID - ID of the linked message used for initial focus.
+ * allReportActions - Array of messages.
+ * fetchNewerReportActions - Function to fetch more messages.
+ * route - Current route, used to reset states on route change.
+ * isLoading - Loading state indicator.
+ * triggerListID - Used to trigger a listID change.
+ * returns {object} An object containing the sliced message array, the pagination function,
+ *                   index of the linked message, and a unique list ID.
+ */
+const usePaginatedReportActionList = (
+    linkedID: string,
+    localAllReportActions: OnyxTypes.ReportAction[],
+    fetchNewerReportActions: (newestReportAction: OnyxTypes.ReportAction) => void,
+    route: RouteProp<CentralPaneNavigatorParamList, typeof SCREENS.REPORT>,
+    isLoading: boolean,
+    triggerListID: boolean,
+) => {
+    // triggerListID is used when navigating to a chat with messages loaded from LHN. Typically, these include thread actions, task actions, etc. Since these messages aren't the latest, we don't maintain their position and instead trigger a recalculation of their positioning in the list.
+    // we don't set currentReportActionID on initial render as linkedID as it should trigger visibleReportActions after linked message was positioned
+    const [currentReportActionID, setCurrentReportActionID] = useState('');
+    const isFirstLinkedActionRender = useRef(true);
+
+    useLayoutEffect(() => {
+        setCurrentReportActionID('');
+    }, [route]);
+
+    const listID = useMemo(() => {
+        isFirstLinkedActionRender.current = true;
+        listIDCount += 1;
+        return listIDCount;
+        // This needs to be triggered with each navigation to a comment. It happens when the route is changed. triggerListID is needed to trigger a listID change after initial mounting.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [route, triggerListID]);
+
+    const index = useMemo(() => {
+        if (!linkedID || isLoading) {
+            return -1;
+        }
+
+        return localAllReportActions.findIndex((obj) => String(obj.reportActionID) === String(isFirstLinkedActionRender.current ? linkedID : currentReportActionID));
+    }, [localAllReportActions, currentReportActionID, linkedID, isLoading]);
+
+    const visibleReportActions = useMemo(() => {
+        if (!linkedID) {
+            return localAllReportActions;
+        }
+        if (isLoading || index === -1) {
+            return [];
+        }
+
+        if (isFirstLinkedActionRender.current) {
+            return localAllReportActions.slice(index, localAllReportActions.length);
+        }
+        const paginationSize = getInitialPaginationSize();
+        const newStartIndex = index >= paginationSize ? index - paginationSize : 0;
```

Comment:
Yes, it works. Originally, I aimed not to oversimplify everything so that it would be more understandable and readable. However, I guess it's still quite readable as it is.
Added, thanks!